pkg/vere/mars.c

```diff
@@ -0,0 +1,288 @@
+{
+  u3_disk* log_u = mar_u->log_u;
+
+  if ( !eve_d ) {
+    eve_d = log_u->dun_d;
+  }
+  else if ( eve_d <= mar_u->dun_d ) {
+    u3l_log("mars: already computed %" PRIu64 "", eve_d);
+    u3l_log("      state=%" PRIu64 ", log=%" PRIu64 "",
+            mar_u->dun_d, log_u->dun_d);
+    return;
+  }
+  else {
+    eve_d = c3_min(eve_d, log_u->dun_d);
+  }
+
+  if ( !mar_u->dun_d ) {
+    c3_w lif_w;
+
+    if ( c3n == u3_disk_read_meta(log_u->mdb_u, 0, 0, &lif_w) ) {
+      fprintf(stderr, "mars: disk read meta fail\r\n");
+      //  XX exit code, cb
+      //
+      exit(1);
+    }
+
+    if ( c3n == _mars_do_boot(mar_u->log_u, lif_w) ) {
+      fprintf(stderr, "mars: boot fail\r\n");
+      //  XX exit code, cb
+      //
+      exit(1);
+    }
+
+    mar_u->sen_d = mar_u->dun_d = lif_w;
+    u3m_save();
+  }
+
+  if ( mar_u->dun_d == log_u->dun_d ) {
+    u3l_log("mars: nothing to do!");
+    return;
+  }
+
+  u3l_log("---------------- playback starting ----------------");
+
+  if ( (1ULL + eve_d) == log_u->dun_d ) {
+    u3l_log("play: event %" PRIu64 "", log_u->dun_d);
+  }
+  else if ( eve_d != log_u->dun_d ) {
+    u3l_log("play: events %" PRIu64 "-%" PRIu64 " of %" PRIu64 "",
+            (c3_d)(1ULL + mar_u->dun_d),
+            eve_d,
+            log_u->dun_d);
+  }
+  else {
+    u3l_log("play: events %" PRIu64 "-%" PRIu64 "",
+            (c3_d)(1ULL + mar_u->dun_d),
+            eve_d);
+  }
+
+  {
+    c3_d pas_d = mar_u->dun_d;  // last snapshot
+    c3_d mem_d = 0;             // last event to meme
+    c3_w try_w = 0;             // [mem_d] retry count
+
+    while ( mar_u->dun_d < eve_d ) {
+      _mars_step_trace(mar_u->dir_c);
+
+      //  XX get batch from args
+      //
+      switch ( _mars_play_batch(mar_u, c3y, 1024) ) {
+        case _play_yes_e: {
+          u3m_reclaim();
+
+          if ( sap_d && ((mar_u->dun_d - pas_d) >= sap_d) ) {
+            u3m_save();
+            pas_d = mar_u->dun_d;
+            u3l_log("play (%" PRIu64 "): save", mar_u->dun_d);
+          }
+          else {
+            u3l_log("play (%" PRIu64 "): done", mar_u->dun_d);
+          }
+        } break;
+
+        case _play_mem_e: {
+          if ( mem_d != mar_u->dun_d ) {
+            mem_d = mar_u->dun_d;
+            try_w = 0;
+          }
+          else if ( 3 == ++try_w ) {
+            fprintf(stderr, "play (%" PRIu64 "): failed\r\n", mar_u->dun_d + 1);
+            u3m_save();
+            //  XX check loom size, suggest --loom X
+            //  XX exit code, cb
+            //
+            u3_disk_exit(log_u);
+            exit(1);
+          }
+
+          //  XX pack before meld?
+          //
+          if ( u3C.wag_w & u3o_auto_meld ) {
+            u3a_print_memory(stderr, "mars: meld: gained", u3u_meld());
+          }
+          else {
+            u3a_print_memory(stderr, "mars: pack: gained", u3m_pack());
+          }
+        } break;
+
+        //  XX handle any specifically?
+        //
+        case _play_int_e:
```

Comment:
A separate message for an interrupted replay would be nice, since it didn't fail so much as become paused